lib/graph/Module.js

```diff
@@ -403,9 +413,48 @@ class Module {
 		if (!config.kind) {
 			config.kind = "project"; // default
 		}
+		for (const key of Object.keys(config)) {
+			this._normalizeConfigValue(config, key);
+		}
 		return config;
 	}
 
+	/**
+	 * Normalizes the config value at object[key]. If the config value is an object / array,
+	 * this method will descend depth-first on its properties / elements. If the config value is a string
+	 * and contains any environment variables, they will be filled in at this point.
+	 *
+	 * @private
+	 * @param {(object|any[])} object An object or array
+	 * @param {(string|number)} key A key of the object or an index of the array
+	 */
+	_normalizeConfigValue(object, key) {
+		let value = object[key];
+		switch (typeof value) {
+		case "string": {
+			value = value.replace(Module._ENVIRONMENT_VARIABLE_REGEX, (substring) => {
+				const envVarName = substring.slice(4);
+				return process.env[envVarName] || substring;
```

Comment:
Hmm, I agree that the current fallback is bad.
We have to consider that the task has no way of knowing that the user might have forgotten something since it doesn't and shouldn't know about the auto-fill. So there are 2 possible outcomes:
1. _empty string_ is a legal value for the configuration key
2. _empty string_ is an illegal value for the configuration key

In case 2, we are fine, the task will (hopefully) throw an error.

But what about case 1?
In case 1 the task may not fail at all and the results may be very different from what the user expects.
Imagine how hard it might be for a user to figure out why their results are all messed up in this case, when seemingly no error occurred.

For this reason, I think it would be best if we take responsibility and throw a reasonable error ourselves instead of delegating it to the task. At the very least we should warn the user imo.

Reply:
I agree, we can't tell whether an empty string is a legal value for a given configuration option.

Are you proposing to raise an error in case an env-var is not set at all, or also in case it is set to an empty string? By now I think I'm more in favor of the first one, since empty strings may be valid configuration options.

I understand that we both agree the name of the environment variable should never be used as a value.

Reply:
> Are you proposing to raise an error in case an env-var is not set at all, ... By now I think I'm more in favor of the first one, since empty strings may be valid configuration options.

Yes, exactly, I think it is the best way to avoid unexpected behavior

> I understand that we both agree the name of the environment variable should never be used as a value.

Yes, I agree

Reply:
> After discussing this with some colleagues, I'm not sure whether raising an expection will work well with some of the expectations from [SAP/ui5-tooling#935](https://github.com/SAP/ui5-tooling/issues/935).
> 
> Especially when environment variables are used in (transitive-)dependencies (e.g. a reuse-library), you would have to make sure to set all those variables when building the root project, as otherwise you will face an error.
> 
>     1. For things like enabling the debug mode of a custom task, this doesn't seem ideal. Surely you don't want to always have to set `DEBUG=` to _explicitly disable the debug mode_.  I'd rather expect the debug mode to be disabled by default, _unless_  `DEBUG=true` is set.
> 
>     2. If a dependency adds an environment variable in a new release, your project's build would break after updating the dependency, until you also provide the new environment variable (which might be something irrelevant for your project)

I agree, environment variables inside your dependencies yamls sound like a hot mess, I will gladly give that up (see comment below). If we do however restrict the feature to the root project, would you then agree that an exception is the way to go? The only way for an exception to pop up in that case is if the user messed up _their own_ config and updates to dependencies should no longer be a concern